shell: fix core/gpu distribution with uneven task counts across nodes

grondo opened the pull request:
This PR is proposed as a fix for #7174.

The fix (or workaround if you prefer) is purposefully applied very narrowly to avoid erroneously reducing the core/gpu affinity in some unforeseen case. That is, the cpuset in the shell is only adjusted if all the following is true:
 - jobspec is v1
 - not an _explicit_ node exclusive allocation
 - nodes were specified in jobspec (i.e. nnnodes != -1)

Similarly the GPU set is only adjusted if all the above is true and:
 - greater than 1 GPU is allocated to this rank

New `gpus_per_slot` and `node_exclusive` members are added to the object provided by `flux_shell_jobspec_info_unpack(3)` since they are required for detection of the above conditions.

grondo commented:
And I just realized that the mpibind plugin sets `cpu-affinity=off` so these changes will have no effect with mpibind. A similar fix will have to go into mpibind itself I guess.

ryanday36 commented:
> Another approach to this PR would be to move the adjustment of the `cores` and `gpus` strings into the shell itself so that the values returned to plugins are the adjusted values.
> 
> I guess that makes a kind of sense since it would fix all plugins that use these values.

I'll ask Edgar to look at this as well, but an approach that makes the plugins just work sounds good to me.

grondo commented:
Thanks. I think I'll go ahead and make the change to the shell's internal copy of "cores" and "gpus", but will change the current debug message that cores/gpus is being adjusted to a warning. Then I'll test again with mpibind.

I'll set WIP on this PR until I get it reworked.

grondo commented:
I've force pushed a new version of this PR that fixes up the core/gpu assignments in the shell proper, so that `flux_shell_rank_info_unpack(3)` reflects the update core and gpu sets. This allows mpibind to just work:

```console
$ FLUX_SHELL_RC_PATH=/etc/flux/shell/lua.d flux run -N2 -n3 -o mpibind=v true
0.088s: flux-shell[1]:  WARN: corona82: adjusted cores from 0-1 to 0
0.094s: flux-shell[1]: mpibind: 
mpibind: task   0 nths  1 gpus  cpus 0
0.089s: flux-shell[0]: mpibind: 
mpibind: task   0 nths  1 gpus  cpus 0
mpibind: task   1 nths  1 gpus  cpus 1
```

The fix is still only applied under the narrow circumstances listed previously.  The debug message is now updated to a warning (only emitted on affected ranks) so users will know that core/gpu sets on the affected nodes will not match R.
